wknml/nml_generation.py

```diff
@@ -0,0 +1,182 @@
+from wknml import NMLParameters, Group, Edge, Node, Tree, NML, Branchpoint, Comment
+import networkx as nx
+import numpy as np
+
+import logging
+import colorsys
+from typing import Tuple, List, Dict, Union, Any
+
+
+logger = logging.getLogger(__name__)
+
+
+def random_color_rgba():
+    # https://stackoverflow.com/a/43437435/783758
+
+    h, s, l = np.random.random(), 0.5 + np.random.random() / 2.0, 0.4 + np.random.random() / 5.0
+    r, g, b = colorsys.hls_to_rgb(h, l, s)
+    return (r, g, b, 1)
+
+
+def discard_children_hierarchy(groups: List[Group]) -> List[Group]:
+    groups_without_hierarchy = []
+    for group in groups:
+        children = discard_children_hierarchy(group.children)
+        groups_without_hierarchy.append(Group(id=group.id, name=group.name, children=[]))
+        groups_without_hierarchy.extend(children)
+    return groups_without_hierarchy
+
+
+def globalize_tree_ids(group_dict: Dict[str, List[nx.Graph]]):
+    current_id = 1
+    for tree_group in group_dict.values():
+        for tree in tree_group:
+            tree.graph["id"] = current_id
+            current_id += 1
+
+
+def globalize_node_ids(group_dict: Dict[str, List[nx.Graph]]):
+    current_id = 1
+    for tree_group in group_dict.values():
+        for tree_index in range(len(tree_group)):
+            tree = tree_group[tree_index]
+            new_tree = nx.Graph(**tree.graph)
+            edge_mapping_dict = {}
+            for node in tree.node:
+                old_id = tree.nodes[node]["id"]
+                tree.nodes[node]["id"] = current_id
+                edge_mapping_dict[old_id] = current_id
+                new_tree.add_node(current_id, **tree.nodes[node])
+
+                current_id += 1
+            new_edges = []
+            for edge in tree.edges:
+                new_edges.append((edge_mapping_dict[edge[0]], edge_mapping_dict[edge[1]]))
+
+            new_tree.add_edges_from(new_edges)
+            tree_group[tree_index] = new_tree
+
+
+def generate_nml(tree_dict: Union[List[nx.Graph], Dict[str, List[nx.Graph]]], globalize_ids: bool = True, parameters: Dict[str, Any] = {}) -> NML:
+
+  if type(tree_dict) is not dict:
+    tree_dict = ["main_group", tree_dict]
+
+    # todo ensure graph attributes and globalize tree ids and maybe group ids
```

Comment:
What does this mean? :)

Reply:
Whoopsie `{"main_group": tree_dict}` is the right syntax.
I use it to always have a dict after that point. thus I do not need to write the same code for a list and a dict.